torch/_jit_internal.py

```diff
@@ -226,6 +228,68 @@ def can_compile_class(cls):
     return all(has_code)
 
 
+def get_type_hint_captures(fn):
+    """
+    Get a dictionary containing type resolution mappings necessary to resolve types
+    for the literal annotations on 'fn'. These are not considered to be closed-over by fn
+    and must be obtained separately.
+
+    Arguments:
+        fn: A callable.
+    Returns:
+        A Dict[str, Any] containing a mapping from the literal annotations used on
+        fn to the Python objects they refer to.
+    """
+    signature = inspect.signature(fn)
+
+    # First, get the literal type annotations from the function declaration
+    # by source inspection. This accounts for the case in which aliases are used
+    # to annotate the arguments (e.g device_t = torch.device, and then d: device_t).
+    src = inspect.getsource(fn)
+
+    # frontend.py cannot be used here because it includes _jit_internal, so use ast instead.
+    a = ast.parse(dedent(src))
+    if len(a.body) != 1 or not isinstance(a.body[0], ast.FunctionDef):
+        raise RuntimeError(f"Expected {fn} to be a function")
+    f = a.body[0]
+
+    # This function converts ast.Name and ast.Attribute nodes to strings.
+    # The latter is how annotations like torch.device are represented; the Attribute
+    # node contains the value "device" and a reference to an ast.Name node containing "torch".
+    def get_annotation_str(annotation):
+        if isinstance(annotation, ast.Name):
+            return annotation.id
+        elif isinstance(annotation, ast.Attribute):
+            return '.'.join([get_annotation_str(annotation.value), annotation.attr])
+        elif isinstance(annotation, ast.Subscript):
+            return f"{annotation.value}[{get_annotation_str(annotation.slice.value)}]"  # type: ignore
+        elif isinstance(annotation, ast.Tuple):
+            return ','.join([get_annotation_str(elt) for elt in annotation.elts])
+        elif isinstance(annotation, ast.Constant) or isinstance(annotation, ast.NameConstant):
+            return f"{annotation.value}"
+
+        raise RuntimeError(f"Unexpected node type: {type(annotation)}")
```

Comment:
Could you print `fn` as well? When I first encountered the error `RuntimeError: Unexpected node type: <class '_ast.Str'>
` (discussed in another comment) I have no clue where it comes from in a large model.

Reply:
Updated the diff to return `None` for unsupported nodes instead of throwing this error.